### PR TITLE
[improve][ml] Avoid to wrap `CompletionException` in `ManagedLedgerException`

### PR DESCRIPTION
**Pull request by mattisonchao**
### Motivation

When I was investigating a problem, I found a bad exception stack trace. Ii made the exception harder to handle and to do `instanceOf`.

image

### Modifications

- Using the static method to help create `ManagedLedgerException` and `MetaStoreException`.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerException.java [hunk 1]
```diff
@@ -29,46 +32,61 @@ public ManagedLedgerException(String msg) {
...
 
-    public static ManagedLedgerException getManagedLedgerException(Throwable e) {
-        if (e instanceof ManagedLedgerException) {
-            return (ManagedLedgerException) e;
+    /**
+     * Returns the encoding of the given (non-null) exception as a wrapped ManagedLedgerException
+     * unless it is one already. Since we use a lot of CompletableFuture, it will unwrap the completion
+     * exception to help get to the original exception.
+     */
+    public static ManagedLedgerException getManagedLedgerException(@Nonnull Throwable ex) {
```
Comment (nicoloboschi) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerException.java:48]: what about making this method generic in `FutureUtil` ? 

maybe something like
```
@SneakyThrows
    public static <T extends Throwable> T wrapException(Class<T> toClass, Throwable ex) {
        final Throwable rc = FutureUtil.unwrapCompletionException(ex);
        if (rc.getClass().isAssignableFrom(toClass)) {
            return (T) rc;
        }
        return toClass.getConstructor(Throwable.class).newInstance(rc);
    }
```

Comment (eolivelli) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerException.java:48]: `return toClass.getConstructor(Throwable.class).newInstance(rc);`
This implies that it works only with Exception types that has that constructor.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerException.java [hunk 2]
```diff
@@ -29,46 +32,61 @@ public ManagedLedgerException(String msg) {
...
+        if (rc instanceof ManagedLedgerException) {
+            return (ManagedLedgerException) rc;
         }
-        return new ManagedLedgerException(e);
+        return new ManagedLedgerException(rc);
     }
 
     public static class MetaStoreException extends ManagedLedgerException {
-        public MetaStoreException(Throwable t) {
+        protected MetaStoreException(Throwable t) {
```
Comment (eolivelli) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerException.java:57]: this is a public API change

**Review by eolivelli — CHANGES_REQUESTED** [review]

**Comment by eolivelli** [comment]
please note that even if this kind or refactoring reduce the tech debt, we are adding many risks.
because due to the nature of how CompletableFuture works it is possible that in many points of the pulsar code (and in Protocol handler and other plugins) the code expects a CompletionException and if we drop one level of exceptions the code may misbehave (from unhandled exceptions to NPEs)

I wouldn't go this way.

We cannot risk the stability of the system.

If you found a case in which we didn't handle correctly a completion exception then we should take a look to the specific point

**Pull request closed without merging**